NIFI-7865 amqp$header is splitted in the wrong way for "," and "}"

sedadgn opened the pull request:
<!--
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

This PR introduces 2 new properties for the ConsumeAMQP processor: escape.comma and remove.curly.braces.
And one new property for PublishAMQP: unescape.comma

This allows to configure the processors to use escaping for commas and to consistently not use curly braces in the amqp$header attribute.

The default values ensure backwards compatibility.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

sedadgn commented on nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/StringUtils.java:
```diff
@@ -510,4 +514,27 @@ public static String toTitleCase(String input) {
+                // special char
+                result.append(escapeChar);
+            }
+            result.append(curChar);
```
I moved in ConsumeAmqp.java.Because only this class use the method.

kevdoran reviewed (approved):
Thanks for making those suggested changes @sedadgn! +1 from me. @ottobackwards do you agree that this is ready? If so, I will merge.

nandorsoma reviewed (commented):
Hey @sedadgn!

Thank you for your contribution. I've left few comments in the code, though I have a general question too.

I tried to search what kind of headers could fit into the headers property. Is there a common list of them or something like that. Unfortunately I didn't find any. The closest information was that developers often misuse that property, because by design you should store information about the message itself in that field. So a negative example is that if you would like to publish information about a http request, the http header should be in the message, but developers often put that info into the headers property. With that in mind I wondering what could be a real value set of that field and how likely you would have commas in the values. Isn't it an edge case that normally would be better to discourage?

sedadgn commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/PublishAMQP.java:
```diff
@@ -99,6 +101,15 @@
             .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
             .build();
 
+    static final PropertyDescriptor UNESCAPE_COMMA = new PropertyDescriptor.Builder()
```
in PublishAMQP According to this parameter, I unescape the escaped value from ConsumeAMQP. I changed the description. I hope it  is more understandable.

sedadgn commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/ConsumeAMQP.java:
```diff
@@ -192,6 +220,48 @@ private void addAttribute(final Map<String, String> attributes, final String att
+        }
+        if (escapeComma && removeCurlyBraces) {
+            return headers.keySet().stream()
+                    .map(key -> key + "=" + escapeString(headers.get(key).toString()))
```
You are right. I changed the parameter names to ESCAPE_COMMA_VALUE_IN_HEADER and UNESCAPE_COMMA_VALUE_IN_HEADER.

sedadgn commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/ConsumeAMQP.java:
```diff
@@ -192,6 +220,48 @@ private void addAttribute(final Map<String, String> attributes, final String att
+                    .collect(Collectors.joining(", ", "{", "}"));
+        } else if (removeCurlyBraces) {
+            String headerString = headers.toString();
+            return headerString.substring(1, headerString.length() - 1);
```
You are right. I added the control for first and last characters

nandorsoma commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/ConsumeAMQP.java:
```diff
@@ -192,6 +221,54 @@ private void addAttribute(final Map<String, String> attributes, final String att
+                    .collect(Collectors.joining(", ", "{", "}"));
+        } else if (removeCurlyBraces) {
+            String headerString = headers.toString();
+            if(headerString.startsWith("{")){
```
Thank you for changing this. I hope it's not nitpicking but I think it would be better to check for ending and closing characters together. This way it's much more consistent and you can also spare a substring call by using the original one. So, something like:
```
if (headerString.startsWith("{") && headerString.endsWith("}"))  {
    return headerString.substring(1, headerString.length() - 1);        
```
Another option would be to construct the string here like in line 229 without escaping. This way it would be consistent code wise that we are only relying on the Map.toString() in the original/compatible mode.

nandorsoma reviewed (commented):
Hey @sedadgn!
I've found one thing among the new changes after that I think we can wrap the things up.

Another thing. During the 2. review, an alternative solution came to my mind. Couldn't we spare the escaping part by making the separator itself configurable? That one feels a bit more robust for me. There is an example for that: https://nifi.apache.org/docs/nifi-docs/components/nifi-docs/components/org.apache.nifi/nifi-poi-nar/1.9.0/org.apache.nifi.processors.poi.ConvertExcelToCSVProcessor/index.html
I understand that the PR has been open for more than a month now so I don't ask you to change the current behavior - unless you strongly agree -,  just wanted to share that idea.

sedadgn commented:
> Hey @sedadgn! I've found one thing among the new changes after that I think we can wrap the things up.
> 
> Another thing. During the 2. review, an alternative solution came to my mind. Couldn't we spare the escaping part by making the separator itself configurable? That one feels a bit more robust for me. There is an example for that: https://nifi.apache.org/docs/nifi-docs/components/nifi-docs/components/org.apache.nifi/nifi-poi-nar/1.9.0/org.apache.nifi.processors.poi.ConvertExcelToCSVProcessor/index.html I understand that the PR has been open for more than a month now so I don't ask you to change the current behavior - unless you strongly agree -, just wanted to share that idea.

Hello @nandorsoma 
I think the solution is better. Strongly I agree :). I will implement. But I will need some time.
Thank you

sedadgn commented:
> Hello @nandorsoma

Thank you for your suggestion.  I implemented your suggestion. Can you review again.

kevdoran commented on nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/StringUtils.java:
```diff
@@ -28,6 +28,8 @@
 
     public static final String EMPTY = "";
+    //With this regular expression, commas without "\"-escape in string can be found.
+    public static final String REGEX_COMMA_WITHOUT_ESCAPE = "(?<!\\\\),";
```
Is this still used with the new approach? If not, let's remove it.

kevdoran commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/pom.xml:
```diff
@@ -48,6 +50,16 @@ language governing permissions and limitations under the License. -->
             <version>1.15.0-SNAPSHOT</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-text</artifactId>
+            <version>${common-text.version}</version>
+        </dependency>
```
I don't see where this is used. Is this added `commons-text` dependency still needed?

kevdoran commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/PublishAMQP.java:
```diff
@@ -224,7 +238,7 @@ private void updateBuilderFromAttribute(final FlowFile flowFile, final String at
      * {@link AMQPUtils#validateAMQPTimestampProperty}
      */
-    private BasicProperties extractAmqpPropertiesFromFlowFile(FlowFile flowFile) {
+    private BasicProperties extractAmqpPropertiesFromFlowFile(FlowFile flowFile,Character escapeComma) {
```
Is the `escapeComma` parameter name a leftover from the previous approach? I think this is intended to be called something like `headerSeparator`.

kevdoran reviewed (commented):
Hi @sedadgn, thanks again for this contribution. I re-reviewed based on the latest approach and have some feedback. If you are able to make another round of updates, I think this is close to being able to merge. Thanks!

nandorsoma commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/test/java/org/apache/nifi/amqp/processors/ConsumeAMQPTest.java:
```diff
@@ -161,6 +165,138 @@ public void validateSuccessfulConsumeAndTransferToSuccess() throws Exception {
+            successFF.assertAttributeEquals("amqp$routingKey", "key1");
+            successFF.assertAttributeEquals("amqp$exchange", "myExchange");
+            String headers = successFF.getAttribute("amqp$headers");
+            Map<String, String> properties = Splitter.on(",").withKeyValueSeparator("=").split(headers.substring(1,headers.length()-1));
```
I think you can hardcode the expectation here, it would be more visible what is the expected result. I understand that if get back the initial map, then the header is valid but in the case of that processor the output is a String therefore you can assert that. What do you think?

sedadgn replied:
I can not. Because there are two elements in the Hashmap. I don't know in what order these elements will come.

For HashMap "This class makes no guarantees as to the order of the map; in particular, it does not guarantee that the order will remain constant over time."

nandorsoma replied:
I see, it makes sense!

nandorsoma commented on nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:
```diff
@@ -980,4 +980,35 @@ public ValidationResult validate(final String subject, final String value, final
+                    .build();
+        }
+
+        if (input.isEmpty()) {
```
I think you can remove that if, because when the input is empty it will produce the same message that you generate in case of `input.length() != 1`.

nandorsoma commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/pom.xml:
```diff
@@ -21,6 +21,8 @@ language governing permissions and limitations under the License. -->
 
     <properties>
         <amqp-client.version>5.8.0</amqp-client.version>
+        <common-text.version>1.8</common-text.version>
```
Is common-text is used somewhere? I don't find it.

sedadgn replied:
> Hey @sedadgn! Thank you for implementing the alternative approach. I've found few things, can you check them? None of them is a big thing so I also think we are really close to merging. One slight thing. There are a lot of inconsistently placed empty lines, missing spaces. They are not considered blocker issues in the community but can I ask you to clean them? I'd appreciate if you could do.

@nandorsoma 
Thank you for your review. I wrote few comment. I made few changing according your comment. 
Best Regards

nandorsoma reviewed (commented):
Hey @sedadgn!
Thank you for implementing the alternative approach. I've found few things, can you check them? None of them is a big thing so I also think we are really close to merging. One slight thing. There are a lot of inconsistently placed empty lines, missing spaces. They are not considered blocker issues in the community but can I ask you to clean them? I'd appreciate if you could do.

sedadgn commented on nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/PublishAMQP.java:
```diff
@@ -240,20 +254,20 @@ private BasicProperties extractAmqpPropertiesFromFlowFile(FlowFile flowFile) {
-    private Map<String, Object> validateAMQPHeaderProperty(String amqpPropValue) {
-        String[] strEntries = amqpPropValue.split(",");
+    private Map<String, Object> validateAMQPHeaderProperty(String amqpPropValue,Character splitValue) {
+        String[] strEntries = amqpPropValue.split(Pattern.quote(String.valueOf(splitValue)));
```
Pattern.quote create the escaped version of the regex representing. 
For Example '|'. 
Also you can look at this. 
https://stackoverflow.com/questions/10796160/splitting-a-java-string-by-the-pipe-symbol-using-split

nandorsoma replied:
I see, it makes sense. Thanks for including the example, I've learned something new. :)

nandorsoma reviewed (approved):
Thanks for working through the feedback @sedadgn, the current version looks good.

What do you think @kevdoran and @ottobackwards?

sedadgn commented:
> Thanks for working through the feedback @sedadgn, the current version looks good.
> 
> What do you think @kevdoran and @ottobackwards?

Hello @nandorsoma, @ottobackwards  and @kevdoran,
Thank you for your feedback and reviews.
We need this feature in our project. If everything works for you, can you merge?
Thank you

nandorsoma commented:
> > Thanks for working through the feedback @sedadgn, the current version looks good.
> > What do you think @kevdoran and @ottobackwards?
> 
> Hello @nandorsoma, @ottobackwards and @kevdoran, Thank you for your feedback and reviews. We need this feature in our project. If everything works for you, can you merge? Thank you

Hey @sedadgn !
Unfortunately I haven't got permission to merge, therefore we need to wait for a committer to do that. Will try to ping someone!

kevdoran commented:
I should be able to do a final round of review on this today or tomorrow, and assuming everything looks good, can merge.

sedadgn commented:
> I should be able to do a final round of review on this today or tomorrow, and assuming everything looks good, can merge.

Hi Kevin, have you had a chance to review? I wanted to remind. 
Thank you.

kevdoran commented on nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:
```diff
@@ -980,4 +980,25 @@ public ValidationResult validate(final String subject, final String value, final
         }
     }
+
+    public static final Validator SINGLE_CHAR_VALIDATOR = (subject, input, context) -> {
```
I've never worked in this file, but at a glance it seems the intent it to keep all the static validators together in one section at the top of the file, so would you mind relocating this?

sedadgn replied:
I moved the method

kevdoran reviewed (commented):
I'm not convinced that adding guava is necessary just for the split/join use case (I was ok with the previous logic, although perhaps it could be refactored out into a dedicated joinMap/splitMap helper method), but if you do want to stick with guava, it looks to me like most other NARs that bundle it call that out in the NOTICE file in the `nifi-*-bundle/nifi-*-nar/src/main/resources/META-INF/` directory, so that would probably be required in this case as well.

Everything else looks good to me aside from one minor comment about code organization in StandardValidators.

sedadgn commented:
> I'm not convinced that adding guava is necessary just for the split/join use case (I was ok with the previous logic, although perhaps it could be refactored out into a dedicated joinMap/splitMap helper method), but if you do want to stick with guava, it looks to me like most other NARs that bundle it call that out in the NOTICE file in the `nifi-*-bundle/nifi-*-nar/src/main/resources/META-INF/` directory, so that would probably be required in this case as well.
> 
> Everything else looks good to me aside from one minor comment about code organization in StandardValidators.

Hello @kevdoran ,
Thank you very much for your review.
I removed guava and added 2 methods for splitMap/joinMap.

kevdoran reviewed (approved):
Thanks @sedadgn! +1, will merge to `main`
